File: C1.UWP.Word/VB/WordSamples/WordUtils.vb
Imports C1.Xaml.Word.Objects
Imports C1.Xaml.Word
Imports Windows.UI.Xaml
Imports Windows.UI.Popups
Imports Windows.UI
Imports Windows.Storage.Streams
Imports Windows.Storage.Pickers
Imports Windows.Storage
Imports Windows.Foundation
Imports Windows.System
Imports System.Threading.Tasks
Imports System.Text
Imports System.Linq
Imports System.IO
Imports System.Collections.Generic
Imports System

Module WordUtils

    <Extension()>
    Public Function RenderParagraph(ByVal doc As C1WordDocument, text As String, font As Font, rcPage As Rect, rc As Rect, outline As Boolean, linkTarget As Boolean) As Rect
        ' if it won't fit this page, do a page break
        rc.Height = doc.MeasureString(text, font, rc.Width).Height
        If rc.Bottom > rcPage.Bottom Then
            doc.PageBreak()
            rc.Y = rcPage.Top
        End If

        ' draw the string
        doc.DrawString(text, font, Colors.Black, rc)

        ' add headings to outline
        If outline Then
            doc.DrawLine(Colors.Black, rc.X, rc.Y, rc.Right, rc.Y)
        End If

        If linkTarget Then
        End If

        ' update rectangle for next time
        rc = Offset(rc, 0, rc.Height)
        Return rc
    End Function

    <Extension()>
    Public Function RenderParagraph(ByVal doc As C1WordDocument, text As String, font As Font, rcPage As Rect, rc As Rect, outline As Boolean) As Rect
        Return RenderParagraph(doc, text, font, rcPage, rc, outline, False)
    End Function

    <Extension()>
    Public Function RenderParagraph(ByVal doc As C1WordDocument, text As String, font As Font, rcPage As Rect, rc As Rect) As Rect
        Return RenderParagraph(doc, text, font, rcPage, rc, False, False)
    End Function

    Dim _pageCurrent As Integer = 1

    <Extension()>
    Public Function CurrentPage(ByVal doc As C1WordDocument) As Integer
        Return _pageCurrent
    End Function
    <Extension()>
    Public Function CurrentPage(ByVal doc As C1WordDocument, page As Integer) As Integer
        _pageCount = Math.Max(_pageCount, page)
        _pageCurrent = page
        Return _pageCurrent
    End Function
    Dim _pageCount As Integer = 1
    <Extension()>
    Public Function PageCount(ByVal doc As C1WordDocument) As Integer
        Return _pageCount
    End Function

    <Extension()>
    Public Function PageRectangle(ByVal doc As C1WordDocument) As Rect
        Return PageRectangle(doc, New Thickness(doc.LeftMargin, doc.TopMargin, doc.RightMargin, doc.BottomMargin))
    End Function
    <Extension()>
    Public Function PageRectangle(ByVal doc As C1WordDocument, pageMargins As Thickness) As Rect
        Dim sz As Size = doc.PageSize
        Dim left As Double = Math.Min(sz.Width, pageMargins.Left)
        Dim top As Double = Math.Min(sz.Height, pageMargins.Top)
        Dim width As Double = Math.Max(0, sz.Width - (pageMargins.Left + pageMargins.Right))
        Dim height As Double = Math.Max(0, sz.Height - (pageMargins.Top + pageMargins.Bottom))
        Return New Rect(left, top, width, height)
    End Function

    <Extension()>
    Public Sub SetDocumentInfo(ByVal doc As C1WordDocument, title As String, Optional graphicFooter As Boolean = False)
        ' set document info
        Dim di As IC1WordInfo = doc.Info
        di.Author = Strings.DocumentAuthor
        di.Subject = Strings.DocumentSubject
        di.Title = title

        ' footer font
        Dim font As New Font("Arial", 8, RtfFontStyle.Bold)
        Dim fmt As New StringFormat()
        fmt.Alignment = HorizontalAlignment.Right
        fmt.LineAlignment = VerticalAlignment.Bottom

        ' render footers
        If graphicFooter Then
            ' this reopens each page and adds content to them (now we know the page count).
            Dim page As Integer = 0
            While page < doc.PageCount()
                doc.CurrentPage(page)
                Dim text As String = String.Format(Strings.Documentfooter, di.Title, page + 1, doc.PageCount())
                doc.DrawString(text, font, Colors.DarkGray, WordUtils.Inflate(doc.PageRectangle(), -72, -36), fmt)
                page += 1
            End While
        Else
            ' standard footer
            Dim text As String = String.Format(Strings.Documentfooter, di.Title, "|", "|")
            Dim paragraph As New RtfParagraph(doc.CurrentSection.Footer)
            paragraph.Alignment = RtfHorizontalAlignment.Right
            Dim count As Integer = 0
            For Each part As String In text.Split("|"c)
                If Not String.IsNullOrEmpty(part) Then
                    paragraph.Add(New RtfString(part))
                End If
                Select Case count
                    Case 0
                        paragraph.Add(New RtfPageField())
                        Exit Select
                    Case 1
                        paragraph.Add(New RtfNumPagesField())
                        Exit Select
                End Select
                count += 1
            Next
            doc.CurrentSection.Footer.Add(paragraph)
        End If
    End Sub

    <Extension()>
    Public Async Sub Save(ByVal doc As C1WordDocument)
        Dim picker As New FileSavePicker()
        picker.FileTypeChoices.Add(Strings.FileTypeChoicesTip, New List(Of String)() From {".docx"})
        picker.FileTypeChoices.Add(Strings.AlternateFileTypeChoicesTip, New List(Of String)() From {".rtf"})
        picker.DefaultFileExtension = ".docx"
        picker.SuggestedStartLocation = PickerLocationId.DocumentsLibrary
        Dim file As StorageFile = Await picker.PickSaveFileAsync()
        If file IsNot Nothing Then
            Await doc.SaveAsync(file, If(file.Name.ToLower().EndsWith(".rtf"), FileFormat.Rtf, FileFormat.OpenXml))
            Dim dlg As New MessageDialog(Strings.SaveLocationTip + " " + file.Path, "WordSamples")
            dlg.Commands.Add(New UICommand("Open", New UICommandInvokedHandler(Async Function(args)
                                                                                   ' to open the created file (using file extension)
                                                                                   Dim success As Boolean = Await Launcher.LaunchFileAsync(file)
                                                                                   Return New UICommand()
                                                                               End Function)))
            dlg.Commands.Add(New UICommand("Cancel"))
            dlg.CancelCommandIndex = 2
            Await dlg.ShowAsync()
        End If
    End Sub

    <Extension()>
    Public Async Function SaveToStream(ByVal doc As C1WordDocument) As Task(Of MemoryStream)
        Dim ms As New MemoryStream()
        Using imas As New InMemoryRandomAccessStream()
            Await doc.SaveAsync(imas.AsStream(), FileFormat.OpenXml)
        End Using
        Return ms
    End Function

    <Extension()>
    Public Function Inflate(ByVal rc As Rect, dx As Double, dy As Double) As Rect
        rc.X -= dx
        rc.Y -= dy
        rc.Width += 2 * dx
        rc.Height += 2 * dy
        Return rc
    End Function

    <Extension()>
    Public Function Offset(ByVal rc As Rect, dx As Double, dy As Double) As Rect
        rc.X += dx
        rc.Y += dy
        Return rc
    End Function

End Module

File: C1.UWP.Word/CS/WordSamples/WordUtils.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Windows.System;
using Windows.Foundation;
using Windows.Storage;
using Windows.Storage.Pickers;
using Windows.Storage.Streams;
using Windows.UI;
using Windows.UI.Popups;
using Windows.UI.Xaml;

using C1.Xaml.Word;
using C1.Xaml.Word.Objects;

namespace WordSamples
{
    /// <summary>
    /// Utility class on top of C1.Xaml.Word.
    /// </summary>
    public static class WordUtils
    {
        // ***********************************************************************
        // C1WordDocuments for measure a paragraph
        // ***********************************************************************

        //public static double MeasureText(this C1WordDocument doc, string text, Font font, Rect rcPage, double height, RtfHorizontalAlignment align = RtfHorizontalAlignment.Undefined)
        //{
        //    // if it won't fit this page, do a page break
        //    var sf = new StringFormat();
        //    switch (align)
        //    {
        //        case RtfHorizontalAlignment.Center:
        //            sf.Alignment = HorizontalAlignment.Center;
        //            break;
        //        case RtfHorizontalAlignment.Justify:
        //            sf.Alignment = HorizontalAlignment.Stretch;
        //            break;
        //        case RtfHorizontalAlignment.Right:
        //            sf.Alignment = HorizontalAlignment.Right;
        //            break;
        //    }
        //    //sf.FormatFlags |= StringFormatFlags.
        //    rc.Height = doc.MeasureString(text, font, rcPage.Width, sf).Height;
        //    if (rc.Bottom > rcPage.Bottom)
        //    {
        //        doc.PageBreak();
        //        rc.Y = rcPage.Top;
        //    }

        //    // add the paragraph
        //    doc.AddParagraph(text, font, clr, align);
        //    //doc.DrawString(text, font, Colors.Black, rc);

        //    // add headings to outline
        //    if (outline)
        //    {
        //        // top line
        //        var paragraph = (RtfParagraph)doc.Current;
        //        paragraph.TopBorderColor = clr;
        //        paragraph.TopBorderStyle = RtfBorderStyle.Single;
        //        paragraph.TopBorderWidth = 1f;

        //        //doc.DrawLine(Colors.Black, rc.X, rc.Y, rc.Right, rc.Y);
        //        doc.AddBookmark(text);
        //    }

        //    // add link target
        //    if (linkTarget)
        //    {
        //        //doc.AddLink(text);
        //    }

        //    // update rectangle for next time
        //    rc = Offset(rc, 0, rc.Height);
        //    return rc;
        //}

        // ***********************************************************************
        // C1WordDocuments for Rect
        // ***********************************************************************

        // measure a paragraph, skip a page if it won't fit, render it into a rectangle,
        // and update the rectangle for the next paragraph.
        // 
        // optionally mark the paragraph as an outline entry and as a link target.
        //
        // this routine will not break a paragraph across pages. for that, see the Text Flow sample.
        //
        public static Rect RenderParagraph(this C1WordDocument doc, string text, Font font, Rect rcPage, Rect rc, bool outline, bool linkTarget)
        {
            // if it won't fit this page, do a page break
            rc.Height = doc.MeasureString(text, font, rc.Width).Height;
            if (rc.Bottom > rcPage.Bottom)
            {
                doc.PageBreak();
                rc.Y = rcPage.Top;
            }

            // draw the string
            doc.DrawString(text, font, Colors.Black, rc);

            // show bounds (to check word wrapping)
            //var p = Pen.GetPen(Colors.Orange);
            //doc.DrawRectangle(p, rc);

            // add headings to outline
            if (outline)
            {
                doc.DrawLine(Colors.Black, rc.X, rc.Y, rc.Right, rc.Y);
                //doc.AddBookmark(text, 0, rc.Y);
            }

            // add link target
            if (linkTarget)
            {
                //doc.AddTarget(text, rc);
            }

            // update rectangle for next time
            rc = Offset(rc, 0, rc.Height);
            return rc;
        }

        public static Rect RenderParagraph(this C1WordDocument doc, string text, Font font, Rect rcPage, Rect rc, bool outline)
        {
            return RenderParagraph(doc, text, font, rcPage, rc, outline, false);
        }

        public static Rect RenderParagraph(this C1WordDocument doc, string text, Font font, Rect rcPage, Rect rc)
        {
            return RenderParagraph(doc, text, font, rcPage, rc, false, false);
        }

        static int _pageCurrent = 1;
        public static int CurrentPage(this C1WordDocument doc)
        {
            return _pageCurrent;
        }
        public static int CurrentPage(this C1WordDocument doc, int page)
        {
            _pageCount = Math.Max(_pageCount, page);
            _pageCurrent = page;
            return _pageCurrent;
        }
        static int _pageCount = 1;
        public static int PageCount(this C1WordDocument doc)
        {
            return _pageCount;
        }

        public static Rect PageRectangle(this C1WordDocument doc)
        {
            return PageRectangle(doc, new Thickness(doc.LeftMargin, doc.TopMargin, doc.RightMargin, doc.BottomMargin));
        }
        public static Rect PageRectangle(this C1WordDocument doc, Thickness pageMargins)
        {
            var sz = doc.PageSize;
            double left = Math.Min(sz.Width, pageMargins.Left);
            double top = Math.Min(sz.Height, pageMargins.Top);
            double width = Math.Max(0, sz.Width - (pageMargins.Left + pageMargins.Right));
            double height = Math.Max(0, sz.Height - (pageMargins.Top + pageMargins.Bottom));
            return new Rect(left, top, width, height);
        }

        public static void SetDocumentInfo(this C1WordDocument doc, string title, bool graphicFooter = false)
        {
            // set document info
            var di = doc.Info;
            di.Author = Strings.DocumentAuthor;
            di.Subject = Strings.DocumentSubject;
            di.Title = title;

            // footer font
            var font = new Font("Arial", 8, RtfFontStyle.Bold);
            var fmt = new StringFormat();
            fmt.Alignment = HorizontalAlignment.Right;
            fmt.LineAlignment = VerticalAlignment.Bottom;

            // render footers
            if (graphicFooter)
            {
                // this reopens each page and adds content to them (now we know the page count).
                for (int page = 0; page < doc.PageCount(); page++)
                {
                    doc.CurrentPage(page);
                    var text = string.Format(Strings.Documentfooter,
                        di.Title,
                        page + 1,
                        doc.PageCount());
                    doc.DrawString(
                        text,
                        font,
                        Colors.DarkGray,
                        WordUtils.Inflate(doc.PageRectangle(), -72, -36),
                        fmt);
                }
            }
            else
            {
                // standard footer
                var text = string.Format(Strings.Documentfooter, di.Title, "|", "|");
                var paragraph = new RtfParagraph(doc.CurrentSection.Footer);
                paragraph.Alignment = RtfHorizontalAlignment.Right;
                int count = 0;
                foreach (var part in text.Split('|'))
                {
                    if (!string.IsNullOrEmpty(part))
                    {
                        paragraph.Add(new RtfString(part));
                    }
                    switch (count)
                    {
                        case 0:
                            paragraph.Add(new RtfPageField());
                            break;
                        case 1:
                            paragraph.Add(new RtfNumPagesField());
                            break;
                    }
                    count++;
                }
                doc.CurrentSection.Footer.Add(paragraph);
            }
        }

        public static async void Save(this C1WordDocument doc)
        {
            FileSavePicker picker = new FileSavePicker();
            picker.FileTypeChoices.Add(Strings.FileTypeChoicesTip, new List<string>() { ".docx" });
            picker.FileTypeChoices.Add(Strings.AlternateFileTypeChoicesTip, new List<string>() { ".rtf" });
            picker.DefaultFileExtension = ".docx";
            picker.SuggestedStartLocation = PickerLocationId.DocumentsLibrary;
            StorageFile file = await picker.PickSaveFileAsync();
            if (file != null)
            {
                await doc.SaveAsync(file, file.Name.ToLower().EndsWith(".rtf") ? FileFormat.Rtf : FileFormat.OpenXml);

                MessageDialog dlg = new MessageDialog(Strings.SaveLocationTip + " " + file.Path, "WordSamples");
                dlg.Commands.Add(new UICommand("Open", new UICommandInvokedHandler((args) =>
                {
                    // to open the created file (using file extension)
                    var success = Launcher.LaunchFileAsync(file);
                })));
                dlg.Commands.Add(new UICommand("Cancel"));
                dlg.CancelCommandIndex = 2;
                await dlg.ShowAsync();
            }
        }

        public static async Task<MemoryStream> SaveToStream(this C1WordDocument doc)
        {
            var ms = new MemoryStream();
            using (var imas = new InMemoryRandomAccessStream())
            {
                await doc.SaveAsync(imas.AsStream(), FileFormat.OpenXml);
            }
            return ms;
        }

        // ***********************************************************************
        // Extension methods for Rect
        // ***********************************************************************

        public static Rect Inflate(this Rect rc, double dx, double dy)
        {
            rc.X -= dx;
            rc.Y -= dy;
            rc.Width += 2 * dx;
            rc.Height += 2 * dy;
            return rc;
        }

        public static Rect Offset(this Rect rc, double dx, double dy)
        {
            rc.X += dx;
            rc.Y += dy;
            return rc;
        }

    }
}
